annotation-processor/src/main/java/org/hibernate/validator/ap/checks/annotationparameters/AnnotationParametersGroupSequenceCheck.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.ap.checks.annotationparameters;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.AnnotationValue;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Elements;
+import javax.lang.model.util.Types;
+
+import org.hibernate.validator.ap.checks.ConstraintCheckIssue;
+import org.hibernate.validator.ap.util.AnnotationApiHelper;
+import org.hibernate.validator.ap.util.CollectionHelper;
+
+/**
+ * Checks that the GroupSequence definition is valid.
+ * <ul>
+ * <li>the class list contains only interface</li>
+ * <li>the defined group sequence is expandable (no cyclic definition)</li>
+ * <li>the class list contains the hosting bean class (for default group sequence re-definition)</li>
+ * </ul>
+ *
+ * @author Marko Bekhta
+ */
+public class AnnotationParametersGroupSequenceCheck extends AnnotationParametersAbstractCheck {
+
+	private Types typeUtils;
+
+	private Elements elementUtils;
+
+	public AnnotationParametersGroupSequenceCheck(AnnotationApiHelper annotationApiHelper, Elements elementUtils, Types typeUtils) {
+		super( annotationApiHelper, "javax.validation.GroupSequence" );
+		this.typeUtils = typeUtils;
+		this.elementUtils = elementUtils;
+	}
+
+	@Override
+	protected Set<ConstraintCheckIssue> doCheck(Element element, AnnotationMirror annotation) {
+		List<? extends AnnotationValue> annotationValue = annotationApiHelper.getAnnotationArrayValue( annotation, "value" );
+
+		TypeElement annotatedElement = (TypeElement) element;
+
+		boolean classForRedefiningGroupSequencePresent = false;
+
+		for ( AnnotationValue value : annotationValue ) {
+			// 1. The class list contains only interface
+			TypeMirror typeMirror = (TypeMirror) value.getValue();
+			boolean isClassAndSameAsAnnotatedElement = isClassForRedefiningGroupSequence( annotatedElement, typeMirror );
+
+			if ( !annotationApiHelper.isInterface( typeMirror ) && !isClassAndSameAsAnnotatedElement ) {
+				return CollectionHelper.asSet( ConstraintCheckIssue.error(
+						element, annotation, "INVALID_GROUP_SEQUENCE_VALUE_NOT_INTERFACES_ANNOTATION_PARAMETERS"
+				) );
+			}
+			// 2. the defined group sequence is expandable (no cyclic definition)
+			if ( isFromHierarchy( annotatedElement, typeMirror ) && !isClassAndSameAsAnnotatedElement ) {
```

Comment:
Do we agree that you don't detect cyclic definitions defined with `@GroupSequence`s on interfaces defining a group sequence? AFAICS, you only check the hierarchy, not the `@GroupSequence`s.

I'm thinking about something like:
```java
@GroupSequence(Group2.class)
public interface Group1 {
}

@GroupSequence(Group1.class)
public interface Group2 {
}
```

Not sure we can do anything about it but this is probably the most common case of cyclic definition (it can of course be far more complicated to detect than this very simple example).